### PR TITLE
SONARJAVA-1056 Rule: S2912 "indexOf" checks should use a start position

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(
+      Tree.Kind.GREATER_THAN, Tree.Kind.GREATER_THAN_OR_EQUAL_TO,
+      Tree.Kind.LESS_THAN, Tree.Kind.LESS_THAN_OR_EQUAL_TO,
+      Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    checkBinaryExpression(((BinaryExpressionTree) tree));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java]: NewMAJORRemove those useless parentheses.[rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.UNDERSTANDABILITY)
+@SqaleConstantRemediation("5min")
+public class IndexOfStartPositionCheck extends SubscriptionBaseVisitor {
+
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+  private static final MethodInvocationMatcher INDEX_OF_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(JAVA_LANG_STRING).name("indexOf").addParameter(JAVA_LANG_STRING);
+
+  public void scanFile(JavaFileScannerContext context) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java]: NewMAJORAdd the "@Override" annotation above this method signature[rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)
NewMINORRemove this method to simply inherit it.[rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1185)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 major
* 1 minor

Details:
* [Add the "@Override" annotation above this method signature](https://github.com/SonarSource/sonar-java/blob/b37194663170891f394f77df15985e4959ed045b/java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java#L54)[rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)
* [Remove this method to simply inherit it.](https://github.com/SonarSource/sonar-java/blob/b37194663170891f394f77df15985e4959ed045b/java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java#L54)[rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1185)
* [Remove those useless parentheses.](https://github.com/SonarSource/sonar-java/blob/b37194663170891f394f77df15985e4959ed045b/java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java#L68)[rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java (hunk 2 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java]: useless override ! just remove the method as indicated by sonarTech ! :)

File: java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+  private void checkBinaryExpression(BinaryExpressionTree expressionTree) {
+    ExpressionTree leftOperand = removeParenthesis(expressionTree.leftOperand());
+    ExpressionTree rightOperand = removeParenthesis(expressionTree.rightOperand());
+    if (leftOperand.is(Tree.Kind.METHOD_INVOCATION)) {
+      checkIndexOfInvocation((MethodInvocationTree) leftOperand, rightOperand);
+    } else if (rightOperand.is(Tree.Kind.METHOD_INVOCATION)) {
+      checkIndexOfInvocation((MethodInvocationTree) rightOperand, leftOperand);
+    } else {
+      // ignore
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java]: else is not really needed here.

File: java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      Tree.Kind.LESS_THAN, Tree.Kind.LESS_THAN_OR_EQUAL_TO,
+      Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    checkBinaryExpression(((BinaryExpressionTree) tree));
+  }
+
+  private void checkBinaryExpression(BinaryExpressionTree expressionTree) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/IndexOfStartPositionCheck.java]: This method should probably be inlined.

**Comment by benzonico** [comment]
see comments, otherwise LGTM. Please fix sonartech issues.

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

**Pull request closed without merging**